src/rgw/rgw_lua_utils.h

```diff
@@ -100,45 +100,57 @@ constexpr auto ONE_RETURNVAL    = 1;
 constexpr auto TWO_RETURNVALS   = 2;
 constexpr auto THREE_RETURNVALS = 3;
 constexpr auto FOUR_RETURNVALS  = 4;
-// utility functions to create a metatable
+
+// create_metatable() is a utility functions to create a metatable
 // and tie it to an unnamed table
 //
-// add an __index method to it, to allow reading values
+// it add an __index method to it, to allow reading values
 // if "readonly" parameter is set to "false", it will also add
 // a __newindex method to it, to allow writing values
 // if the "toplevel" parameter is set to "true", it will name the
 // table as well as the metatable, this would allow direct access from
 // the lua script.
+// the name of the metatable will be a fully qualified name, based on the names
+// of the metatables that contain it, separated by dots.
+// it also adds a __len and  __pairs methods for iterable entities.
 //
 // The MetaTable is expected to be a class with the following members:
-// Name (static function returning the unique name of the metatable)
-// TableName (static function returning the unique name of the table - needed only for "toplevel" tables)
 // Type (typename) - the type of the "upvalue" (the type that the meta table represent)
 // IndexClosure (static function return "int" and accept "lua_State*") 
 // NewIndexClosure (static function return "int" and accept "lua_State*") 
+// PairsClosure (static function return "int" and accept "lua_State*") 
+// LenClosure (static function return "int" and accept "lua_State*") 
+//
 // e.g.
 // struct MyStructMetaTable {
-//   static std::string TableName() {
-//     return "MyStruct";
-//   }
 //
 //   using Type = MyStruct;
 //
 //   static int IndexClosure(lua_State* L) {
-//     const auto value = reinterpret_cast<const Type*>(lua_touserdata(L, lua_upvalueindex(FIRST_UPVAL)));
+//     const auto name = tatable_name_upvalue(L);
+//     const auto value = reinterpret_cast<const Type*>(lua_touserdata(L, lua_upvalueindex(SECOND_UPVAL)));
 //     ...
 //   }
 
 //   static int NewIndexClosure(lua_State* L) {
+//     const auto name = table_name_upvalue(L);
 //     auto value = reinterpret_cast<Type*>(lua_touserdata(L, lua_upvalueindex(FIRST_UPVAL)));
 //     ...
 //   }
 // };
 //
+// If the structs inherits from EmptyMetatable then the only function that needs 
+// to be implemented is IndexClosure. The other functions are implemented in 
+// the base class stating that the table is "readonly" and "not-iterable".
 
 template<typename MetaTable, typename... Upvalues>
-void create_metatable(lua_State* L, bool toplevel, Upvalues... upvalues)
+void create_metatable(lua_State* L, const std::string_view parent_name, const std::string_view field_name,
```

Comment:
```suggestion
void create_metatable(lua_State* L, std::string_view parent_name, std::string_view field_name,
```

`const std::string_view` is redundant